base/compiler/ssair/inlining.jl

```diff
@@ -1225,6 +1213,100 @@ function process_simple!(ir::IRCode, idx::Int, state::InliningState, todo::Vecto
     return stmt, sig
 end
 
+"""
+    is_emit_atomicrmw_eligible(f)
+
+Check if a function `f` is an `IntrinsicFunction` that can be handled by `emit_atomicrmw`
+(codegen).
+"""
+is_emit_atomicrmw_eligible(@nospecialize f) =
+    f === Intrinsics.add_int ||
+    f === Intrinsics.sub_int ||
+    f === Intrinsics.and_int ||
+    f === Intrinsics.or_int ||
+    f === Intrinsics.xor_int ||
+    f === Intrinsics.add_float ||
+    f === Intrinsics.sub_float
+
+"""
+    optimize_modifyfield!(inst::Instruction, modifysig::Signature, state::InliningState)
+
+Optimize `modifyfield!` in `inst` in-place.
+
+If `inst` is of form `modifyfield!(struct, field, op, arg2, order)` where `op` is (a method
+wrapping) a simple call of intrinsic, transform `inst[:inst]` to a `:call` to the builtin
+function `modifyfield!`.  This is detected in codegen and lowered to `atomicrmw` instruction
+(see `emit_atomicrmw`).
+
+Otherwise, `inst` is transformed to `:invoke_modify` which is lowered to method invocation
+combined atomic replace in codegen.
+"""
+function optimize_modifyfield!(inst::Instruction, modifysig::Signature, state::InliningState)
+    stmt = inst[:inst]
+    @assert stmt.head === :call
+
+    info = inst[:info]
+    info isa MethodResultPure && (info = info.info)
+    info isa ConstCallInfo && (info = info.call)
+    info isa MethodMatchInfo || return nothing
+    length(info.results) == 1 || return nothing
+    opmatch = info.results[1]::MethodMatch
+    opmatch.fully_covers || return nothing
+    opcase = compileable_specialization(state.et, opmatch, Effects())
+    opcase === nothing && return nothing
+
+    optimize_modifyfield_as_intrinsic!(inst, modifysig, opcase, opmatch, state) &&
+        return nothing
+
+    stmt.head = :invoke_modify
+    pushfirst!(stmt.args, opcase.invoke)
+
+    return nothing
+end
+
+function optimize_modifyfield_as_intrinsic!(
+    inst::Instruction, modifysig::Signature, opcase::InvokeCase, opmatch::MethodMatch,
+    state::InliningState)
+
+    @assert inst[:inst].head === :call
+
+    # Extract argument types for `op(arg1, arg2)` from the signature of
+    # `modifyfield!(struct, field, op, arg2, order)`.
+    fieldarg = modifysig.argtypes[3]
+    fieldarg isa Const || return false
+    field = fieldarg.val
+    field isa Union{Int,Symbol} || return false
+    structtype = widenconst(modifysig.argtypes[2])
+    arg1type = fieldtype(structtype, field)
+    arg2type = widenconst(modifysig.argtypes[5])
+    (isconcretetype(arg1type) && isconcretetype(arg2type) && arg1type == arg2type) ||
+        return false
+    ftype = modifysig.argtypes[4]
+    argtypes = Any[ftype, arg1type, arg2type]
```

Comment:
why does this look complicated, when `abstract_modifyfield!` seems to implement the whole lattice in fewer lines?